File: src/v/model/validation.cc
// Copyright 2020 Redpanda Data, Inc.
//
// Use of this software is governed by the Business Source License
// included in the file licenses/BSL.md
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0

#include "model/validation.h"

#include "base/seastarx.h"
#include "base/vlog.h"
#include "model/errc.h"

#include <seastar/util/log.hh>

namespace model {

std::error_code validate_kafka_topic_name(const model::topic& tpc) {
    static constexpr size_t kafka_max_topic_name_length = 249;
    const auto& name = tpc();
    if (name.empty()) {
        return make_error_code(errc::topic_name_empty);
    }
    if (name == "." || name == "..") {
        return make_error_code(errc::forbidden_topic_name);
    }
    if (name.length() > kafka_max_topic_name_length) {
        return make_error_code(errc::topic_name_len_exceeded);
    }
    for (char c : name) {
        if (!std::isalnum(c) && !(c == '.' || c == '-' || c == '_')) {
            return make_error_code(errc::invalid_topic_name);
        }
    }
    return make_error_code(errc::success);
}

} // namespace model

File: src/go/rpk/pkg/cli/security/role/validation.go
// Copyright 2026 Redpanda Data, Inc.
//
// Use of this software is governed by the Business Source License
// included in the file licenses/BSL.md
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the Apache License, Version 2.0

package role

import (
	"k8s.io/apimachinery/pkg/util/validation"
)

// validateRoleNameForK8s reports whether the role name is a valid DNS-1123
// subdomain. Returns nil if the name is compliant; otherwise returns the
// validation messages from k8s.io/apimachinery so the caller can surface them
// to the user.
//
// The Redpanda Kubernetes operator binds a RedpandaRole CR's role name to its
// metadata.name, which the Kubernetes API server constrains to DNS-1123. Names
// that don't satisfy this constraint cannot be adopted by a RedpandaRole CR
// and will block migration to operator-managed roles.
func validateRoleNameForK8s(name string) []string {
	return validation.IsDNS1123Subdomain(name)
}
